components/automate-ui/src/app/modules/infra-proxy/treetable/component/treetable.component.ts

```diff
@@ -0,0 +1,102 @@
+import { Subject } from 'rxjs';
+import * as O from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+
+@Component({
+  selector: 'app-treetable',
+  templateUrl: './treetable.component.html',
+  styleUrls: ['./treetable.component.scss']
+})
+export class TreetableComponent<T> implements OnInit, OnChanges {
+  @Input() tree: Node<T> | Node<T>[];
+  @Input() options: Options<T> = {};
+  @Output() nodeClicked: Subject<TreeTableNode<T>> = new Subject();
+  private searchableTree: SearchableNode<T>[];
+  private treeTable: TreeTableNode<T>[];
+  displayedColumns: string[];
+  dataSource: MatTableDataSource<TreeTableNode<T>>;
+  constructor(
+    private treeService: TreeService,
+    private validatorService: ValidatorService,
+    private converterService: ConverterService
+  ) { }
+
+  ngOnInit() {
+    this.tree = Array.isArray(this.tree) ? this.tree : [this.tree];
+    this.options = this.parseOptions(defaultOptions);
+    const customOrderValidator =
+    this.validatorService.validateCustomOrder(this.tree[0], this.options.customColumnOrder);
+    if (this.options.customColumnOrder && !customOrderValidator.valid) {
+      throw new Error(`
+        Properties ${customOrderValidator.xor.map(x => `'${x}'`).join(', ')} incorrect or missing in customColumnOrder`
+      );
+    }
+    this.displayedColumns = this.options.customColumnOrder
+      ? this.options.customColumnOrder
+      : this.extractNodeProps(this.tree[0]);
+    this.searchableTree = this.tree.map(t => this.converterService.toSearchableTree(t));
+    const treeTableTree = this.searchableTree.map(st => this.converterService.toTreeTableTree(st));
+    this.treeTable = flatMap(treeTableTree, this.treeService.flatten);
+    this.dataSource = this.generateDataSource();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.tree.isFirstChange()) {
+      return;
+    }
+    this.tree = Array.isArray(this.tree) ? this.tree : [this.tree];
+    this.searchableTree = this.tree.map(t => this.converterService.toSearchableTree(t));
+    const treeTableTree = this.searchableTree.map(st => this.converterService.toTreeTableTree(st));
+    this.treeTable = flatMap(treeTableTree, this.treeService.flatten);
+    this.dataSource = this.generateDataSource();
+  }
+
+  extractNodeProps(tree: Node<T> & { value: { [k: string]: any } }): string[] {
+    return Object.keys(tree.value).filter(x => typeof tree.value[x] !== 'object');
+  }
+
+  generateDataSource(): MatTableDataSource<TreeTableNode<T>> {
+    return new MatTableDataSource(this.treeTable.filter(x => x.isVisible));
+  }
+
+  formatIndentation(node: TreeTableNode<T>, step: number = 5): string {
+    return '&nbsp;'.repeat(node.depth * step);
+  }
+
+  formatElevation(): string {
+    return `mat-elevation-z${this.options.elevation}`;
+  }
+
+  onNodeClick(clickedNode: TreeTableNode<T>): void {
+    clickedNode.isExpanded = !clickedNode.isExpanded;
+    this.treeTable.forEach(el => {
```

Comment:
Please add a comment here as well. This is my guess--is this correct?
```
// A given element `el` is marked visible if every node between it and the root is expanded.
```

Also, wondering why you need to run through `O.fold`: is it possible that a node in the `treeTable` might not be found by `searchById`? I guess I am not at all clear on what you are doing here, since there is a lot going on, with searchableTree, treeService, and treeTable. Please add a couple paragraphs in the code explaining how these work. Thanks!

Reply:
> Please add a comment here as well. This is my guess--is this correct?
> 
> ```
> // A given element `el` is marked visible if every node between it and the root is expanded.
> ```
> 
> Also, wondering why you need to run through `O.fold`: is it possible that a node in the `treeTable` might not be found by `searchById`? I guess I am not at all clear on what you are doing here, since there is a lot going on, with searchableTree, treeService, and treeTable. Please add a couple paragraphs in the code explaining how these work. Thanks!

yes correct it is.